### PR TITLE
Connecting Flutter App to a Database

### LINKED ISSUE #33: Connect the app to a database

### PR DESCRIPTION
**Pull request by MapleSyyrup**
- Connect the app to firebase to save the products in the database

image

image

### CODE REVIEW

File: lib/providers/products_provider.dart [hunk 1]
```diff
@@ -48,9 +51,29 @@ class ProductsProvider with ChangeNotifier {
   Product findById(String id) => _items.firstWhere((prod) => prod.id == id);
 
   /// addProduct is called when the user adds a new product in the shop
-  void addProduct(Product product) {
-    _items.add(product);
-    notifyListeners();
+  Future<void> addProduct(Product product) {
+    const url = 'https://shop-app-4ed38-default-rtdb.firebaseio.com/products.json';
```
Comment (Nei1eveN) [inline lib/providers/products_provider.dart]: kindly add this to Contants

File: lib/providers/products_provider.dart [hunk 2]
```diff
@@ -48,9 +51,29 @@ class ProductsProvider with ChangeNotifier {
...
+    return http.post(
+      url,
+      body: json.encode({
+        'title': product.title,
```
Comment (Nei1eveN) [inline lib/providers/products_provider.dart]: this `product.title` seems repeated in `.then` block, same with `product.description` and others. extract them as variables instead.

File: lib/screens/cart_screen.dart [hunk 3]
```diff
@@ -36,15 +36,15 @@ class CartScreen extends StatelessWidget {
...
+                  TextButton(
                     onPressed: () {
                       Provider.of<Orders>(context, listen: false).addOrder(
                         cart.items.values.toList(),
                         cart.totalAmount,
                       );
                       cart.clear();
                     },
-                    textColor: primaryColor,
+                    // textColor: primaryColor,
```
Comment (Nei1eveN) [inline lib/screens/cart_screen.dart]: remove if no longer needed

File: lib/screens/edit_product_screen.dart [hunk 4]
```diff
@@ -102,6 +103,9 @@ class _EditProductScreenState extends State<EditProductScreen> {
       return;
     }
     _formKey.currentState.save();
+    setState(() {
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart]: expression

File: lib/screens/edit_product_screen.dart [hunk 5]
```diff
@@ -113,10 +117,27 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+    // _editedProduct.id != null
+    //     ? provider.updateProduct(product)
+    //     : provider.addProduct(product).then(
+    //           (value) => Navigator.of(context).pop(),
+    //         );
+    if (_editedProduct.id != null) {
+      provider.updateProduct(product);
+      setState(() {
+        _isLoading = false;
+      });
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart]: convert this to an expression

File: lib/screens/edit_product_screen.dart [hunk 6]
```diff
@@ -113,10 +117,27 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
+      Navigator.of(context).pop();
+    } else {
+      provider.addProduct(product).then((_) {
+        setState(() {
+          _isLoading = false;
+        });
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart]: same here

File: lib/screens/edit_product_screen.dart [hunk 7]
```diff
@@ -176,78 +197,82 @@ class _EditProductScreenState extends State<EditProductScreen> {
...
-                children: [
-                  Container(
-                    width: 100,
-                    height: 100,
-                    margin: const EdgeInsets.only(
-                      top: 4,
-                      right: 10,
+      body: _isLoading
+          ? Center(
+              child: CircularProgressIndicator(),
```
Comment (Nei1eveN) [inline lib/screens/edit_product_screen.dart]: trailing commas

File: lib/screens/edit_product_screen.dart (hunk 6 above)
Comment (Nei1eveN) [reply lib/screens/edit_product_screen.dart]: then perform this setState after the if-else blocks instead 🤔

**Review by Nei1eveN — CHANGES_REQUESTED** [review]

File: lib/providers/products_provider.dart (hunk 2 above)
Comment (MapleSyyrup) [reply lib/providers/products_provider.dart]: It is different. In `json.encode` it is a map while in `.then` block, it is the `Product` class.

**Review by Nei1eveN — APPROVED** [review]

**Pull request merged**